src/activities/ActivityUsageEntity.js

```diff
@@ -187,6 +187,30 @@ export class ActivityUsageEntity extends Entity {
 		await this._setDate(dateValue, Classes.dates.endDate, 'endDate');
 	}
 
+	/**
+	 * Updates start date, due date and end date together to the dates specified
+	 * @param {string} startDate Date string to set as the start date, or empty string to clear the start date
+	 * @param {string} dueDate Date string to set as the due date, or empty string to clear the due date
+	 * @param {string} endDate Date string to set as the end date, or empty string to clear the end date
+	 */
+	async setDates(startDate, dueDate, endDate) {
+		let action;
+		const datesEntity = this._getDateSubEntity('dates');
+		if (datesEntity) {
+			action = datesEntity.getActionByName(Actions.activities.update);
+		}
+
+		if (!action) {
+			return;
+		}
+		const fields = [
+			{ name: 'startDate', value: startDate },
+			{ name: 'dueDate', value: dueDate },
+			{ name: 'endDate', value: endDate }
```

Comment:
Here we are passing all three dates (startDate, dueDate and endDate) from the UI to the backend -- however, is this required to do? Would it be better to just pass the dates that have changed their values and use the backend to source the unchanged dates' values?

Reply:
Hmm, good question. My initial reaction is we should only send the dates that have changed. That tends to be how we have done other PATCH APIs. I think the only reason why we might want to always sends all dates, is if we want to guard against the possibility that someone is working with stale data e.g. someone else has modified the dates whilst the current user has the edit page open and there is the possibility that a different user has modified one of the dates to be inconsistent with the current user's change. Then the current user would get a validation error even though the dates they see on the page are consistent when they clicked save.

But if we allow them to send all their dates even though they have not changed some of them, that means they clobber other users changes even though they didn't explicitly change the field.

Might be a question for Joseph.

In general I think our goal is to not overwrite other changes if the current user didn't make changes, hence we only send changed data.  So my preference would be to not send all the dates and only send ones that have changed in the UI. But it might mean we have to make our validation handling smarter e.g. by refreshing the dates from the backend if we detect a validation error caused by stale data.

Reply:
Confirmed with Joseph and apparently we _do_ want to overwrite saved changes made by any other user. It's what quizzing does as well when editing questions.
In that case we'll keep this as-is, so that the backend always has all the new date values to do the validation.

Reply:
I'd like to double check that because it means it's inconsistent with how we're doing other things. e.g. if the user edits the name, we don't save the instructions too.

Regardless, even if we stick with this approach of always saving all dates, I think we should enhance the `*DateChanged` checks to determine if the date has actually changed rather than just checking for undefined. If we leave it as is it means we are saving all the dates every time we save, which is again inconsistent with how we are doing other things?

Unless that is also what Joseph wants? To save everything every time? But that means we are making lots of extra API calls, and for some things it's harder to do and requires comparison checking on the backend. eg. for attachments.

Reply:
^ Updated so it only saves when at least one of the dates has been changed.